### DISCUSSION TITLE
Esp32-Cam & Microdot

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#16170] [ESP32]
Hi,
I would like to display on a web page the video of my new Esp32-Cam that I just received.
I noticed that 90% of the tutorials are based on Arduino.
I also saw that there were two main firmware:
https://github.com/shariltumin/esp32-cam-micropython-2022, I could not install it and this one:
https://github.com/lemariva/micropython-camera-driver, which works on my card.
To start I wanted to display an image on a web page, and I'm already stuck.
The main.py and foto.jpg files are on the root
main.py
```
def wifi_ap():#adress IP ='192.168.4.10'
    ap = network.WLAN(network.AP_IF)
    ap.active(True)
    ap.ifconfig((ip_ap,subnet_ap,gateway_ap,dns_ap))
    ap.config(essid=ssid, password=password)
    while ap.active() == False:
         pass
    print('Active server')
    print(ap.ifconfig())
    return()
    
wifi_ap()

app = Microdot()
Response.default_content_type = 'text/html'

@app.get('/')
async def index(req):
    return Template('index.html').render()

@app.route('/static/<path:path>')
async def static(request, path):
    if '..' in path:
        return 'Not found', 404
    return send_file('static/' + path)

@app.get('/shutdown')
async def shutdown(request):
    request.app.shutdown()
    return 'The server is shutting down...'

app.run(debug=True)
```
index.html
```
<!DOCTYPE html>
<html lang="fr">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Test camera</title>
  </head>
  <body>
    <h2>CAMERA</h2> 
    <p>Image</p>
      <img src="foto.jpg" width=400 height=341
   alt="Nothing"/>
   </body>
</html>
```
It's really basic but the image is not displayed on the web page, and I have the following message:
```
Active server
('192.168.4.10', '255.255.255.0', '192.168.4.1', '192.168.4.1')
Starting async server on 0.0.0.0:5000...
GET / 200
GET /foto.jpg 404
```
index.html file works fine with Visual Studio Code
What am I missing?
Thanks for your help

**Comment by shariltumin** [comment]
Perhaps these will help:

change,
```
<img src="foto.jpg" width=400 height=341 alt="Nothing"/>
```

to 
```
<img src="/static/foto.jpg" width=400 height=341 alt="Nothing"/>
```

and create a "/static" directory in your flash and move your image file there so that you have a "/static/foto.jpg" file.

**Reply by PatLub85** [reply]
I made the change but it's the same thing

**Reply by PatLub85** [reply]
Sorry, too quick to reply, it works. Microdot creates an index_html.py file that doesn't pick up changes immediately

**Reply by PatLub85** [reply]
Question:
why an image is not detected if it is on the root

**Reply by shariltumin** [reply]
This is because your route
```
@app.route('/static/<path:path>')
async def static(request, path):
    if '..' in path:
        return 'Not found', 404
    return send_file('static/' + path)
```

needs '/static/whateverfilë́'

If you have
```
@app.route('/foto.jpg')
async def static(request, path):
    if '..' in path:
        return 'Not found', 404
    return send_file('foto.jpg')
```

then you can have the image in root. But this is not practical, how about foto1.jpg, foto2.jpg, and so on? You don't want to hardcode all those files.

**Reply by PatLub85** [reply]
ok i understand
Now I would like to move on to the next step:
Can we directly send a photo taken by the ESP WITHOUT saving it in the memory
Something like:
```
def photo():
    camera.init(0,format=camera.JPEG, fb_location=camera.PSRAM)
    camera.framesize(camera.FRAME_VGA)
    buf=camera.capture()
    camera.deinit()
    return buf
@app.get('/')
async def index(req):
    return Template('index.html').render(foto=photo())
```
Index.html
```
{% args foto %}
<!DOCTYPE html>
<html lang="fr">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Test camera</title>

  </head>
  <body>
    <h2>CAMERA</h2> 
    <p>Image</p>
    <img src={{ foto }} width=400 height=341
   alt="Nothing"/>
 
  </body>
</html>

```
It doesn't work, I get the binary code of the image

**Reply by shariltumin** [reply]
Sorry, I don't use Microdot. My firmware do not even include asyncio. You can find some examples from microdot ripo, for example https://github.com/miguelgrinberg/microdot/blob/main/examples/streaming/video_stream.py

**Reply by PatLub85** [reply]
I managed to install your firmwares-20230717 and I will try your webcam

**Reply by shariltumin** [reply]
You may also want to try a more recent firmware by cnadler86 from https://github.com/cnadler86/micropython-camera-API/tree/master

**Reply by PatLub85** [reply]
I modified wifi.py with the SSID and password and I get:
```
>>> execfile('webcam.py')
Camera ready?:  False
Camera ready?:  False
Camera ready?:  False
Camera ready?:  False
Camera ready?:  False
Timeout
```
What is missing?

**Reply by PatLub85** [reply]
I removed the for loop by cam = camera.init()
and I get:
```
PWD: oQ1GduN9
Traceback (most recent call last):
 File "<stdin>", line 166, in <module>
NameError: name 'w' isn't defined
```

line 166: site.ip=w.wlan.ifconfig()[0]

**Reply by shariltumin** [reply]
What type of esp32-cam board and camera sensor are you using?

**Reply by PatLub85** [reply]
it's an AliExpress clone. 
On the other hand I understood my problem, before restarting the program, you have to do a camera.desinit()
Now the server is running but when I log in I get a message saying I am not authorized

**Reply by shariltumin** [reply]
You can try to disable authentication by

```
# set auth
# auth.on=True
auth.on=False  # False: no authentication needed
```

**Reply by PatLub85** [reply]
I went shopping. I'll try tomorrow morning. See you tomorrow

**Reply by PatLub85** [reply]
Hi Shariltumine
Now I can connect but the page that appears on my screen are the following:
console
```
Camera ready?:  True
Camera ready
Waiting ...
Waiting ...
Connected to Wifi_tests
network config: ('192.168.144.25', '255.255.255.0', '192.168.144.1', '192.168.144.1')
Start server 80
Try - http://192.168.144.25
('192.168.144.5', 49170) ['GET', '/']
```
my screen
```
Autentication:
  Login to server with a one-time password.
  http://192.168.144.25/login/<PWD>
  After login, only the authenticated client can connect to server.

  Logout from server.
  http://192.168.144.25/logout
  The client is no longer authenticated. Create new one-time password.
  The password is shown in REPL.
  Any web browser can now login using the password.
..............
  whitebalance=0 # 0:default, 1:sunny, 2:cloudy, 3:office, 4:home
  aelevels=0     # [-2,2] AE Level: Automatic exposure
  aecvalue=0     # [0,1200] AEC Value: Automatic exposure control
  agcgain=0      # [0,30] AGC Gain: Automatic Gain Control
```
???

**Reply by PatLub85** [reply]
It's help(server)  in help.py

**Comment by cnadler86** [comment]
Here is also one firmware with webcam example: [micropython-camera-API](https://github.com/cnadler86/micropython-camera-API)
